File: Helper/search_tool.py
#!/usr/bin/env python3
import os
import sys

def normalize_bool(choice: str) -> bool:
    return choice.strip() in ("1", "y", "Y", "yes", "YES", "Yes")

def contains(haystack: str, needle: str, case_insensitive: bool) -> bool:
    if case_insensitive:
        return needle.lower() in haystack.lower()
    return needle in haystack

def search_by_filename(base_dir: str, keyword: str, case_insensitive: bool):
    results = []
    for root, _, files in os.walk(base_dir):
        for name in files:
            if contains(name, keyword, case_insensitive):
                results.append(os.path.join(root, name))
    return results

def is_text_file(path: str) -> bool:
    # Heuristic: treat as text if it doesn't contain NUL and can be decoded as UTF-8 with replacement
    try:
        with open(path, "rb") as f:
            chunk = f.read(4096)
        if b"\x00" in chunk:
            return False
        return True
    except Exception:
        return False

def file_contains_keyword(path: str, keyword: str, case_insensitive: bool) -> bool:
    # Read line by line to avoid large memory usage
    try:
        with open(path, "r", encoding="utf-8", errors="ignore") as f:
            if case_insensitive:
                key = keyword.lower()
                for line in f:
                    if key in line.lower():
                        return True
            else:
                for line in f:
                    if keyword in line:
                        return True
        return False
    except Exception:
        return False

def search_by_file_content(base_dir: str, keyword: str, case_insensitive: bool):
    results = []
    for root, _, files in os.walk(base_dir):
        for name in files:
            path = os.path.join(root, name)
            if not is_text_file(path):
                continue
            if file_contains_keyword(path, keyword, case_insensitive):
                results.append(path)
    return results

def search_by_foldername(base_dir: str, keyword: str, case_insensitive: bool):
    results = []
    for root, dirs, _ in os.walk(base_dir):
        for d in dirs:
            if contains(d, keyword, case_insensitive):
                results.append(os.path.join(root, d))
    return results

def main():
    base_dir = input("Enter base directory (e.g. /srv/folderA): ").strip()
    if not os.path.isdir(base_dir):
        print(f"ERROR: Directory not found: {base_dir}")
        sys.exit(1)

    print("\nSelect function:")
    print("1) Search file NAME contains keyword")
    print("2) Search file CONTENT contains keyword")
    print("3) Search FOLDER NAME contains keyword")
    mode = input("Enter 1, 2 or 3: ").strip()

    print("\nCase sensitivity:")
    print("1) Case-INSENSITIVE")
    print("2) Case-SENSITIVE")
    case_mode = input("Enter 1 or 2: ").strip()
    case_insensitive = (case_mode == "1")

    keyword = input("\nSearch keyword: ").strip()
    print("\nResult:\n-------------------")

    if mode == "1":
        results = search_by_filename(base_dir, keyword, case_insensitive)
        for p in results:
            print(p)
        print(f"\nSummary\n{base_dir.rstrip('/')}/ : {len(results)} files")
        return

    if mode == "2":
        results = search_by_file_content(base_dir, keyword, case_insensitive)
        for p in results:
            print(p)
        print(f"\nSummary\n{base_dir.rstrip('/')}/ : {len(results)} files")
        return

    if mode == "3":
        results = search_by_foldername(base_dir, keyword, case_insensitive)
        for p in results:
            print(p)
        print(f"\nSummary\n{base_dir.rstrip('/')}/ : {len(results)} folders")
        return

    print("ERROR: Invalid selection. Please enter 1, 2, or 3.")
    sys.exit(1)

if __name__ == "__main__":
    main()

File: Helper/search_tool.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Find;
use Cwd 'abs_path';

sub contains {
    my ($haystack, $needle, $case_insensitive) = @_;
    if ($case_insensitive) {
        return lc($haystack) =~ /\Q@{[lc($needle)]}\E/;
    }
    return $haystack =~ /\Q$needle\E/;
}

sub is_text_file {
    my ($path) = @_;
    return 0 if !-f $path;
    open my $fh, "<", $path or return 0;
    binmode($fh);
    read($fh, my $buf, 4096);
    close $fh;
    return ($buf !~ /\x00/); # simple binary check
}

print "Enter base directory (e.g. /srv/folderA): ";
chomp(my $base_dir = <STDIN>);

die "ERROR: Directory not found: $base_dir\n" if !-d $base_dir;

print "\nSelect function:\n";
print "1) Search file NAME contains keyword\n";
print "2) Search file CONTENT contains keyword\n";
print "3) Search FOLDER NAME contains keyword\n";
print "Enter 1, 2 or 3: ";
chomp(my $mode = <STDIN>);

print "\nCase sensitivity:\n";
print "1) Case-INSENSITIVE\n";
print "2) Case-SENSITIVE\n";
print "Enter 1 or 2: ";
chomp(my $case_mode = <STDIN>);
my $case_insensitive = ($case_mode eq "1") ? 1 : 0;

print "\nSearch keyword: ";
chomp(my $keyword = <STDIN>);

print "\nResult:\n-------------------\n";

my @results;

if ($mode eq "1") {
    find(sub {
        return if !-f $_;
        if (contains($_, $keyword, $case_insensitive)) {
            push @results, $File::Find::name;
        }
    }, $base_dir);

    print "$_\n" for @results;
    print "\nSummary\n" . ($base_dir =~ s{/$}{}r) . "/ : " . scalar(@results) . " files\n";
    exit 0;
}

if ($mode eq "2") {
    find(sub {
        return if !-f $_;
        my $path = $File::Find::name;

        return if !is_text_file($path);

        open my $fh, "<:encoding(UTF-8)", $path or return;
        my $found = 0;

        if ($case_insensitive) {
            my $k = lc($keyword);
            while (my $line = <$fh>) {
                if (index(lc($line), $k) >= 0) { $found = 1; last; }
            }
        } else {
            while (my $line = <$fh>) {
                if (index($line, $keyword) >= 0) { $found = 1; last; }
            }
        }

        close $fh;

        if ($found) {
            push @results, $path;
        }
    }, $base_dir);

    print "$_\n" for @results;
    print "\nSummary\n" . ($base_dir =~ s{/$}{}r) . "/ : " . scalar(@results) . " files\n";
    exit 0;
}

if ($mode eq "3") {
    find(sub {
        return if !-d $_;
        return if $File::Find::name eq $base_dir; # skip base dir itself

        # Compare folder "basename"
        my $folder_name = $_;
        if (contains($folder_name, $keyword, $case_insensitive)) {
            push @results, $File::Find::name;
        }
    }, $base_dir);

    print "$_\n" for @results;
    print "\nSummary\n" . ($base_dir =~ s{/$}{}r) . "/ : " . scalar(@results) . " folders\n";
    exit 0;
}

die "ERROR: Invalid selection. Please enter 1, 2, or 3.\n";